File: DataPipeline/Python/Web/Plain/src/storage/adapters/postgresql.py
import os
import psycopg2
import psycopg2.extras
from storage.warehouse_adapter import DataWarehouseAdapter


class PostgreSQL(DataWarehouseAdapter):
    def __init__(self):
        self.conn = None
        self.connect()

    def connect(self):
        if self.conn and not self.conn.closed:
            return
        self.conn = psycopg2.connect(
            host=os.getenv("DB_HOST", "db"),
            port=int(os.getenv("DB_PORT", "5432")),
            dbname=os.getenv("DB_NAME", "etl_pipeline"),
            user=os.getenv("DB_USER", "postgres"),
            password=os.getenv("DB_PASSWORD", "postgres"),
        )
        self.conn.autocommit = True

    def execute(self, query: str, params: tuple = None) -> list:
        self.connect()
        cur = self.conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        cur.execute(query, params or ())
        if cur.description:
            rows = [dict(r) for r in cur.fetchall()]
        else:
            rows = []
        cur.close()
        return rows

    def bulk_insert(self, table: str, records: list) -> None:
        if not records:
            return
        self.connect()
        cur = self.conn.cursor()
        cols = list(records[0].keys())
        placeholders = ", ".join(["%s"] * len(cols))
        col_names = ", ".join(cols)
        values = [tuple(r.get(c) for c in cols) for r in records]
        cur.executemany(f"INSERT INTO {table} ({col_names}) VALUES ({placeholders})", values)
        cur.close()

    def create_table(self, table: str, schema: dict) -> None:
        self.connect()
        col_defs = ", ".join(f"{name} {dtype}" for name, dtype in schema.items())
        cur = self.conn.cursor()
        cur.execute(f"CREATE TABLE IF NOT EXISTS {table} ({col_defs})")
        cur.close()

    def list_tables(self) -> list:
        self.connect()
        cur = self.conn.cursor()
        cur.execute("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public'")
        tables = [r[0] for r in cur.fetchall()]
        cur.close()
        return tables

    def close(self):
        if self.conn and not self.conn.closed:
            self.conn.close()

File: CloudLocal/aws/localstack/pipeline/Node.js/Web/Plain/src/storage/Adapters/PostgreSQL.js
const { Client } = require('pg');
const DatabaseAdapter = require('../DatabaseAdapter');

class PostgreSQLAdapter extends DatabaseAdapter {
  constructor() {
    super();
    this._client = null;
  }

  async connect() {
    this._client = new Client({
      host: process.env.DB_HOST || 'localhost',
      port: parseInt(process.env.DB_PORT || '5432'),
      database: process.env.DB_NAME || 'pipeline_db',
      user: process.env.DB_USER || 'admin',
      password: process.env.DB_PASSWORD || 'admin',
    });
    await this._client.connect();
    console.log('Connected to PostgreSQL');
  }

  async query(sql, params) {
    if (!this._client) await this.connect();
    const result = await this._client.query(sql, params);
    return result.rows;
  }

  async insert(table, data) {
    const columns = Object.keys(data);
    const values = Object.values(data);
    const placeholders = columns.map((_, i) => `$${i + 1}`).join(', ');
    const columnNames = columns.join(', ');
    const sql = `INSERT INTO ${table} (${columnNames}) VALUES (${placeholders}) RETURNING *`;
    const result = await this.query(sql, values);
    return result[0];
  }

  async update(table, id, data) {
    const columns = Object.keys(data);
    const values = Object.values(data);
    const setClause = columns.map((col, i) => `${col} = $${i + 1}`).join(', ');
    const sql = `UPDATE ${table} SET ${setClause} WHERE id = $${columns.length + 1} RETURNING *`;
    const result = await this.query(sql, [...values, id]);
    return result[0];
  }

  async delete(table, id) {
    const sql = `DELETE FROM ${table} WHERE id = $1`;
    await this.query(sql, [id]);
  }

  async find(table, id) {
    const sql = `SELECT * FROM ${table} WHERE id = $1`;
    const result = await this.query(sql, [id]);
    return result[0] || null;
  }

  async findAll(table) {
    const sql = `SELECT * FROM ${table}`;
    return this.query(sql);
  }

  async close() {
    if (this._client) {
      await this._client.end();
      this._client = null;
    }
  }
}

module.exports = { PostgreSQLAdapter };
